### ISSUE TITLE
'write() failed: Broken pipe' error

### ISSUE THREAD

**Issue by Josef-Ouano** [ideasman42/nerd-dictation#102]
Hello,
I am trying to run nerd-dictation on a Wayland machine.
But I keep getting this error:
> Traceback (most recent call last):
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1952, in <module>
    main()
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1948, in main
    args.func(args)
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1813, in <lambda>
    func=lambda args: main_begin(
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1416, in main_begin
    found_any = text_from_vosk_pipe(
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1207, in text_from_vosk_pipe
    json_text, json_text_partial_prev = rec_handle_fn_wrapper_from_partial_result(json_text_partial_prev)
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1073, in rec_handle_fn_wrapper_from_partial_result
    handle_fn_wrapper(text, True)
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 1031, in handle_fn_wrapper
    handle_fn(len(text_prev) - match, text_curr[match:])
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 198, in simulate_typing_with_ydotool
    run_command_or_exit_on_failure(
  File "/home/ubuntu/Documents/test/./nerd-dictation", line 50, in run_command_or_exit_on_failure
    subprocess.check_output(cmd)
  File "/usr/lib/python3.10/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ydotool', 'type', '--next-delay', '5', '--', 'just']' returned non-zero exit status 2.
write() failed: Broken pipe

The Command I used is the ff:
**./nerd-dictation begin --simulate-input-tool=YDOTOOL --vosk-model-dir=./lib/language_models/model_en**

I've cloned ydotool from the link: https://github.com/ReimuNotMoe/ydotool

I have no idea why this happens, I've never been able to make it work

**Comment by sal-zen** [comment]
Hello!
I've got the same issue on Wayland!

./nerd-dictation begin --simulate-input-tool=YDOTOOL --vosk-model-dir=./model
ydotool: notice: ydotoold backend unavailable (may have latency+delay issues)
terminate called after throwing an instance of 'std::runtime_error'
  what():  failed to open uinput device
Traceback (most recent call last):
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1974, in <module>
    main()
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1970, in main
    args.func(args)
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1835, in <lambda>
    func=lambda args: main_begin(
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1437, in main_begin
    found_any = text_from_vosk_pipe(
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1225, in text_from_vosk_pipe
    json_text, json_text_partial_prev = rec_handle_fn_wrapper_from_partial_result(json_text_partial_prev)
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1084, in rec_handle_fn_wrapper_from_partial_result
    handle_fn_wrapper(text, True)
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 1042, in handle_fn_wrapper
    handle_fn(len(text_prev) - match, text_curr[match:])
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 199, in simulate_typing_with_ydotool
    run_command_or_exit_on_failure(
  File "/home/salim/.config/nerd-dictation/./nerd-dictation", line 50, in run_command_or_exit_on_failure
    subprocess.check_output(cmd)
  File "/usr/lib/python3.10/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ydotool', 'type', '--next-delay', '5', '--', 'hello']' died with <Signals.SIGABRT: 6>.
write() failed: Broken pipe

**Comment by sal-zen** [comment]
I asked chatgpt 4 about the issue and here it's reply:

**My Question:** 
act as an linux expert, and give tell me why nerd-dictation doesn't work in wayland ubutnu even if I use it with ydotool?

**ChatGPT's answer:**
"Nerd-dictation" and "ydotool" are tools that interact directly with the X.org server, which is the underlying technology for the GUI in many Linux distributions. Nerd-dictation uses the X11 server's capabilities for speech recognition to transcribe spoken language into written text, while ydotool is a command-line tool that simulates keyboard input.

However, Wayland is a newer display server protocol that is designed to replace X.org in many Linux distributions. Wayland provides better security by isolating each application's input and output, which means that applications can't interfere with or spy on each other. Unfortunately, this also means that tools that rely on interacting directly with the X.org server, like nerd-dictation and ydotool, may not work properly with Wayland.

The problem with nerd-dictation and ydotool on Wayland is that they rely on the ability to inject input events into the system. In X.org, this is relatively simple because X.org doesn't make a strong distinction between "real" input events (like actual key presses) and "synthetic" input events (like those created by ydotool). But in Wayland, the security model is different. Applications are not allowed to inject input events into the system, because this could be used to control or spy on other applications.

There are potential workarounds for this issue, like using a virtual keyboard protocol, but these require changes to the applications and may not provide a seamless experience. In addition, not all Wayland compositors support these protocols.

So in short, the reason why nerd-dictation and ydotool don't work on Wayland is that Wayland's security model doesn't allow applications to inject input events into the system. This is a feature, not a bug, but it does create problems for certain types of applications. Until these tools are updated to work with Wayland's security model, they will likely continue to have issues on Wayland.

**Comment by ideasman42** [comment]
@sal-zen ChatGPT's answer was obviously wrong with minimal research. `ydotool` doesn't depend on X11.

**Comment by sal-zen** [comment]
@ideasman42 : Thanks for your clarification. Do you have any general hypothesis on why I got the error? I can further explore them and share what I'll find here to solve the problem.
Thank you for your collab.

**Comment by ideasman42** [comment]
@sal-zen can you get ydotool to work in the command line? If so, it shouldn't be a problem to run it from nerd-dictation.

**Comment by sal-zen** [comment]
@ideasman42 I was able to spot the error and now it works fine for me.
I needed to to do these:

1. lunch ydotoold before lunching ydotool in a seperate terminal window. To do this run : sudo ydotoold
2. run nerd-dictation using the command : ./nerd-dictation begin --timeout 5 --simulate-input-tool STDOUT --vosk-model-dir=./model

Now it works fine for me!
Thank you @ideasman42 for you reactivity and your willingness to help. I appreciate it :)

**Comment by AshkanArabim** [comment]
I used `dotool` instead of `ydotool` and it's much easier. no need to start a server or anything.
- to start: `nerd-dictation begin --simulate-input-tool=DOTOOL`
- to top: `nerd-dictation end`
easy!
